README.md
# **JSON API Errors**

JSON API Errors is a collection of popular errors which is format under of JSON API Specification. More detailed in [here](https://jsonapi.org/format/#error-objects)

## Getting Started

### Installation
The easiest way to install `json-api-error` is using NPM. If you have Node.js installed, it is most likely that you have NPM installed as well.

```
$ npm install json-api-error
```

### Usage

At the current implementation, `json-api-error` support the following common errors

- `BadRequestError`
- `UnauthorizedError`
- `MalformedError`
- `ForbiddenError`
- `NotFoundError`
- `MethodNotAllowedError`
- `NotAcceptableError`
- `RequestTimeoutError`
- `UnsupportedMediaTypeError`
- `InternalError`
- `NotImplementedError`
- `BadGatewayError`
- `ServiceUnavailableError`
- `GatewayTimeoutError`

### Example

### Error usage with simple detailed message

```js
import { BadRequestError } from 'json-api-error';

throw new BadRequestError('Request is invalid');
```

### Error usage with options

```
import { BadRequestError } from 'json-api-error';

throw new BadRequestError({
    id: 'BadRequestError',
    code: 'BadRequestError',
    title: 'BadRequestError',
    detail: 'The #/userName must be number'
}));
```

### Error usage with customized your JsonApiError

```js
import JsonApiError from 'json-api-error';

throw new JsonApiError({
    id: 'RequestEntityTooLargeError',
    status: '413',
    code: 'RequestEntityTooLargeError',
    title: 'RequestEntityTooLargeError',
    detail: 'Request Entity Too Large'
  });
```

### AggregateJsonApiError(errors, status)

```js
import { AggregateJsonApiError } from 'json-api-error';

throw new AggregateJsonApiError([
  new BadRequestError('Something went wrong'),
  new BadRequestError('Request is valid')
], 400);
```

### Express Middleware - JSON API Error Handler usage

`json-api-error` also ships a Express middleware to handle these JSON API Errors. This middleware will catch JSON API Error and return response for your end-user as the following example format (including the status response):

```js
{
  errors: [{
    id: 'NotFoundError',
    status: '404',
    code: 'NotFoundError',
    title: 'NotFoundError',
    detail: 'Resource was not found'
  }]
}
```

To use this, configure your Express app


```js
import { jsonApiErrorHandler } from 'json-api-error/middlewares';


app.use(jsonApiErrorHandler);

```

**Note:** For best practice, please place `jsonApiErrorHandler` below utilized middlewares.

### Options

| Properties  | Detail   | Type  |  Default Value  |
|---|---|---|---|
| id  | a unique identifier for this particular occurrence of the problem  | string  |  equal to error name |
| links  | a links object, more detailed in [here](https://jsonapi.org/format/#error-objects) |  object | N/A |
| status | the HTTP status code applicable to this problem, expressed as a string value  |  string | equal to HTTP code |
| code |  an application-specific error code, expressed as a string value | string  | equal to error name |
| title | a short, human-readable summary of the problem  | string  |  equal to error name |
| detail | a human-readable explanation specific to this occurrence of the problem  | string  |  equal to error name |
| source |  an object containing references to the source of the error, more detailed in [here](https://jsonapi.org/format/#error-objects) |  object | N/A |
| meta | a meta object containing non-standard meta-information about the error.  | object  |  N/A |

## LICENSE

MIT


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]

### Added

- Include typings when publishing to npm

### Changed

- Bump various build dependencies
- Migrate to Jest 28

## [1.1.0] - 06-05-2020

### Added

- Add `AggregateJsonApiError`

## [1.0.6] - 29-10-2019

### Added

- Add `MalformedError`

## [1.0.5] - 29-10-2019

### Changed

- Extract `dist` into package
- Exclude ``dist` into package

## [1.0.4] - 23-10-2019

### Fixed

- Fixed simple detail does not work

### Added

- JSON API Error Handler Middleware for Express

## [1.0.3] - 21-10-2019

### Fixed

- Fixed merging input in constructor with static variable

## [1.0.2] - 08-10-2019

### Changed

- Enable partial options in constructing available Json API Error.

## [1.0.1] - 01-10-2019

### Changed

- Update documentation in `README.md`

### Removed

- Removed `src` & `travis.yml` to be published

## [1.0.0] - 27-09-2019

### Added

- Initial release

[1.1.0]: https://github.com/GeminiWind/JsonAPIErrors/compare/v1.0.6...v1.1.0
[1.0.6]: https://github.com/GeminiWind/JsonAPIErrors/compare/v1.0.5...v1.0.6
[1.0.5]: https://github.com/GeminiWind/JsonAPIErrors/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/GeminiWind/JsonAPIErrors/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/GeminiWind/JsonAPIErrors/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/GeminiWind/JsonAPIErrors/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/GeminiWind/JsonAPIErrors/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/GeminiWind/JsonAPIErrors/releases/tag/v1.0.0


tests/index.test.ts
import {
  default as JsonApiError,
  BadRequestError,
  MalformedError,
  UnauthorizedError,
  ForbiddenError,
  NotFoundError,
  MethodNotAllowedError,
  NotAcceptableError,
  RequestTimeoutError,
  UnsupportedMediaTypeError,
  InternalError,
  NotImplementedError,
  ServiceUnavailableError,
  GatewayTimeoutError
} from '../src/';

describe('errors', () => {
  describe('BadRequestError', () => {
    it('throw BadRequestError', () => {
      expect(() => {
        throw new BadRequestError('Request is invalid.');
      }).toThrow(BadRequestError);
    });

    it('throw BadRequestError with options', () => {
      expect(() => {
        throw new BadRequestError({
          detail: 'Request is invalid'
        });
      }).toThrow(BadRequestError);
    });
  });

  describe('UnauthorizedError', () => {
    it('throw UnauthorizedError', () => {
      expect(() => {
        throw new UnauthorizedError('Unauthorized');
      }).toThrow(UnauthorizedError);
    });

    it('throw UnauthorizedError with options', () => {
      expect(() => {
        throw new UnauthorizedError({
          detail: 'Unauthorized'
        });
      }).toThrow(UnauthorizedError);
    });
  });

  describe('MalformedError', () => {
    it('throw MalformedError', () => {
      expect(() => {
        throw new MalformedError('Error in reading malformed JSON');
      }).toThrow(MalformedError);
    });

    it('throw UnauthorizedError with options', () => {
      expect(() => {
        throw new MalformedError({
          detail: 'Error in reading malformed JSON'
        });
      }).toThrow(MalformedError);
    });
  });


  describe('ForbiddenError', () => {
    it('throw ForbiddenError', () => {
      expect(() => {
        throw new ForbiddenError('Forbidden');
      }).toThrow(ForbiddenError);
    });

    it('throw ForbiddenError with options', () => {
      expect(() => {
        throw new ForbiddenError({
          detail: 'Forbidden'
        });
      }).toThrow(ForbiddenError);
    });
  });

  describe('NotFoundError', () => {
    it('throw NotFoundError', () => {
      expect(() => {
        throw new NotFoundError('Record was not found');
      }).toThrow(NotFoundError);
    });

    it('throw NotFoundError with options', () => {
      expect(() => {
        throw new NotFoundError({
          detail: 'Record was not found'
        });
      }).toThrow(NotFoundError);
    });
  });

  describe('MethodNotAllowedError', () => {
    it('throw MethodNotAllowedError', () => {
      expect(() => {
        throw new MethodNotAllowedError('Method GET does not allowed');
      }).toThrow(MethodNotAllowedError);
    });

    it('throw MethodNotAllowedError with options', () => {
      expect(() => {
        throw new MethodNotAllowedError({
          detail: 'Method GET does not allowed'
        });
      }).toThrow(MethodNotAllowedError);
    });
  });

  describe('NotAcceptableError', () => {
    it('throw NotAcceptableError', () => {
      expect(() => {
        throw new NotAcceptableError('Unacceptable');
      }).toThrow(NotAcceptableError);
    });

    it('throw NotAcceptableError with options', () => {
      expect(() => {
        throw new NotAcceptableError({
          detail: 'Unacceptable'
        });
      }).toThrow(NotAcceptableError);
    });
  });

  describe('RequestTimeoutError', () => {
    it('throw RequestTimeoutError', () => {
      expect(() => {
        throw new RequestTimeoutError('Request timeout');
      }).toThrow(RequestTimeoutError);
    });

    it('throw RequestTimeoutError with options', () => {
      expect(() => {
        throw new RequestTimeoutError({
          detail: 'Request timeout'
        });
      }).toThrow(RequestTimeoutError);
    });
  });

  describe('UnsupportedMediaTypeError', () => {
    it('throw UnsupportedMediaTypeError', () => {
      expect(() => {
        throw new UnsupportedMediaTypeError('Content-type: application/json does not supported');
      }).toThrow(UnsupportedMediaTypeError);
    });

    it('throw UnsupportedMediaTypeError with options', () => {
      expect(() => {
        throw new UnsupportedMediaTypeError({
          detail: 'Content-type: application/json does not supported'
        });
      }).toThrow(UnsupportedMediaTypeError);
    });
  });

  describe('InternalError', () => {
    it('throw InternalError', () => {
      expect(() => {
        throw new InternalError('Internal Server Error.');
      }).toThrow(InternalError);
    });

    it('throw InternalError with options', () => {
      expect(() => {
        throw new InternalError({
          detail: 'Internal Server Error.'
        });
      }).toThrow(InternalError);
    });
  });

  describe('NotImplementedError', () => {
    it('throw NotImplementedError', () => {
      expect(() => {
        throw new NotImplementedError('Not Implemented');
      }).toThrow(NotImplementedError);
    });

    it('throw NotImplementedError with options', () => {
      expect(() => {
        throw new NotImplementedError({
          detail: 'Not Implemented'
        });
      }).toThrow(NotImplementedError);
    });
  });

  describe('ServiceUnavailableError', () => {
    it('throw ServiceUnavailableError', () => {
      expect(() => {
        throw new ServiceUnavailableError('Service is unavailable');
      }).toThrow(ServiceUnavailableError);
    });

    it('throw ServiceUnavailableError with options', () => {
      expect(() => {
        throw new ServiceUnavailableError({
          detail: 'Service is unavailable'
        });
      }).toThrow(ServiceUnavailableError);
    });
  });

  describe('GatewayTimeoutError', () => {
    it('throw GatewayTimeoutError', () => {
      expect(() => {
        throw new GatewayTimeoutError('Gateway timeout');
      }).toThrow(GatewayTimeoutError);
    });

    it('throw GatewayTimeoutError with options', () => {
      expect(() => {
        throw new GatewayTimeoutError({
          detail: 'Gateway timeout'
        });
      }).toThrow(GatewayTimeoutError);
    });
  });

  describe('JSonApiError', () => {
    it('should throw customized JsonApiError if JsonApiError option is valid', () => {
      expect(() => {
        throw new JsonApiError({
          id: 'RequestEntityTooLargeError',
          status: '413',
          code: 'RequestEntityTooLargeError',
          title: 'RequestEntityTooLargeError',
          detail: 'Request Entity Too Large'
        });
      }).toThrow(JsonApiError);
    });

    it('throw error if JsonApiError options is invalid ', () => {
      expect(() => {
        throw new JsonApiError(33333);
      }).toThrow(TypeError);
    });

    it('toJSON', () => {
      const func = () => {
        throw new JsonApiError({
          id: 'RequestEntityTooLargeError',
          status: '413',
          code: 'RequestEntityTooLargeError',
          title: 'RequestEntityTooLargeError',
          detail: 'Request Entity Too Large'
        });
      };

      let error;

      try {
        func();
      } catch (e) {
        error = e;
      }

      const jsonError = error.toJSON();

      expect(error).toBeInstanceOf(JsonApiError);

      expect(jsonError).toHaveProperty('id', 'RequestEntityTooLargeError');
      expect(jsonError).toHaveProperty('status', '413');
      expect(jsonError).toHaveProperty('code', 'RequestEntityTooLargeError');
      expect(jsonError).toHaveProperty('title', 'RequestEntityTooLargeError');
      expect(jsonError).toHaveProperty('detail', 'Request Entity Too Large');
    });
  });
});


tests/AggregateJsonApiError.test.ts
import AggregateJsonApiError from '../src/AggregateJsonApiError';
import { BadRequestError } from '../src/';

describe('AggregateJsonApiError', () => {
  it('throw BadRequestError', () => {
    expect(() => {
      throw new AggregateJsonApiError([
        new BadRequestError('Data is required.'),
        new BadRequestError('Request is invalid.'),
      ], 400);
    }).toThrow(AggregateJsonApiError);
  });
});


tests/middlewares/jsonApiErrorHandler.test.ts
import { jsonApiErrorHandler } from '../../src/middlewares';
import { NotFoundError, BadRequestError, AggregateJsonApiError } from '../../src';

describe('jsonApiErrorHandler', () => {
  it('should return response in the JSON API Error format if error is instance of JsonApiError', () => {
    const err = new NotFoundError('Resource was not found');

    const res = {
      send:  jest.fn().mockImplementation(function () { return this; }),
      status: jest.fn().mockImplementation(function () { return this; })
    };

    const next = jest.fn();

    jsonApiErrorHandler(err, undefined, res, next);

    expect(res.status).toHaveBeenCalledWith(404);
    expect(res.send).toHaveBeenCalledWith({
      errors: [{
        id: 'NotFoundError',
        status: '404',
        code: 'NotFoundError',
        title: 'NotFoundError',
        detail: 'Resource was not found'
      }]
    });
  });

  it('should return response in the JSON API Error format if error is instance of AggregateJsonApiError', () => {
    const err = new AggregateJsonApiError([
      new BadRequestError('Data is required'),
      new BadRequestError('Record is not found'),
    ], 400);

    const res = {
      send:  jest.fn().mockImplementation(function () { return this; }),
      status: jest.fn().mockImplementation(function () { return this; })
    };

    const next = jest.fn();

    jsonApiErrorHandler(err, undefined, res, next);

    expect(res.status).toHaveBeenCalledWith(400);
    expect(res.send).toHaveBeenCalledWith({
      errors: [{
        id: 'BadRequestError',
        status: '400',
        code: 'BadRequestError',
        title: 'BadRequestError',
        detail: 'Data is required'
      }, {
        id: 'BadRequestError',
        status: '400',
        code: 'BadRequestError',
        title: 'BadRequestError',
        detail: 'Record is not found'
      }]
    });
  });

  it('should forward error to later processing if error is not instance of JsonApiError', () => {
    const err = new Error('This is not JsonApiError');

    const res = {
      send:  jest.fn().mockImplementation(function () { return this; }),
      status: jest.fn().mockImplementation(function () { return this; })
    };

    const next = jest.fn();

    jsonApiErrorHandler(err, undefined, res, next);

    expect(res.status).toHaveBeenCalledTimes(0);
    expect(res.send).toHaveBeenCalledTimes(0);

    expect(next).toHaveBeenCalledWith(err);
  });
});


src/JsonApiError.ts
import { CustomError } from 'ts-custom-error';

interface ILink {
  about: string;
}

interface ISource {
  pointer: string;
  parameter: string;
}

interface IMeta {
  [key: string]: any;
}
interface IJsonApiError {
  id: string;
  link?: ILink;
  status: string;
  code: string;
  title: string;
  detail: string;
  source?: ISource;
  meta?: IMeta;
}

interface IJsonApiErrorOptionalField {
  id?: string;
  link?: ILink;
  code?: string;
  title?: string;
  detail?: string;
  source?: ISource;
  meta?: IMeta;
}

class JsonApiError extends CustomError {
  static id: string;
  static link: ILink;
  static status: string;
  static code: string;
  static title: string;
  static detail: string;
  static source: ISource;
  static meta: IMeta;

  private _error;

  static readonly allowedProps: string[] = [
    'id',
    'link',
    'status',
    'code',
    'title',
    'detail',
    'source',
    'meta'
  ];

  constructor(input: any);
  constructor(input: IJsonApiErrorOptionalField);
  constructor(input: IJsonApiError) {
    let opts: any = {};

    const inputType = typeof input;

    switch (inputType) {
    case 'string':
      opts.detail = input;
      break;
    case 'object':
      opts = Object.keys(input).reduce((acc, k) => {
        if (k && input[k] && JsonApiError.allowedProps.includes(k)) {
          acc[k] = input[k];
        }

        return acc;
      }, {});
      break;
    default:
      throw new TypeError(`Could not parse ${input} as JsonApiError options`);
    }

    super(opts.detail);

    this._error = JsonApiError.allowedProps.reduce(
      (error: object, propName: string): any => {
        if (opts[propName]) {
          error[propName] = opts[propName];
        } else if (this.constructor[propName]) {
          error[propName] = this.constructor[propName];
        }

        return error;
      },
      {}
    );
  }

  toJSON() {
    return this._error;
  }
}

export default JsonApiError;


src/index.ts
import JsonApiError from './JsonApiError';
import AggregateJsonApiError from './AggregateJsonApiError';

export class BadRequestError extends JsonApiError {
  static id = 'BadRequestError';
  static status = '400';
  static code = 'BadRequestError';
  static title = 'BadRequestError';
  static detail = '[BadRequestError] - BadRequestError';
}

export class UnauthorizedError extends JsonApiError {
  static id = 'UnauthorizedError';
  static status = '401';
  static code = 'UnauthorizedError';
  static title = 'UnauthorizedError';
  static detail = '[UnauthorizedError] - UnauthorizedError';
}

export class ForbiddenError extends JsonApiError {
  static id = 'ForbiddenError';
  static status = '403';
  static code = 'ForbiddenError';
  static title = 'ForbiddenError';
  static detail = '[ForbiddenError] - ForbiddenError';
}

export class NotFoundError extends JsonApiError {
  static id = 'NotFoundError';
  static status = '404';
  static code = 'NotFoundError';
  static title = 'NotFoundError';
  static detail = '[NotFoundError] - NotFoundError';
}

export class MethodNotAllowedError extends JsonApiError {
  static id = 'MethodNotAllowedError';
  static status = '405';
  static code = 'MethodNotAllowedError';
  static title = 'MethodNotAllowedError';
  static detail = '[MethodNotAllowedError] - MethodNotAllowedError';
}

export class NotAcceptableError extends JsonApiError {
  static id = 'NotAcceptableError';
  static status = '406';
  static code = 'NotAcceptableError';
  static title = 'NotAcceptableError';
  static detail = '[NotAcceptableError] - NotAcceptableError';
}

export class RequestTimeoutError extends JsonApiError {
  static id = 'RequestTimeoutError';
  static status = '408';
  static code = 'RequestTimeoutError';
  static title = 'RequestTimeoutError';
  static detail = '[RequestTimeoutError] - RequestTimeoutError';
}

export class UnsupportedMediaTypeError extends JsonApiError {
  static id = 'UnsupportedMediaTypeError';
  static status = '415';
  static code = 'UnsupportedMediaTypeError';
  static title = 'UnsupportedMediaTypeError';
  static detail = '[UnsupportedMediaTypeError] - UnsupportedMediaTypeError';
}

export class InternalError extends JsonApiError {
  static id = 'InternalError';
  static status = '500';
  static code = 'InternalError';
  static title = 'InternalError';
  static detail = '[InternalError] - InternalError';
}

export class MalformedError extends JsonApiError {
  static id = 'MalformedError';
  static status = '400';
  static code = 'MalformedError';
  static title = 'MalformedError';
  static detail = '[MalformedError] - Error in reading malformed JSON';
}

export class NotImplementedError extends JsonApiError {
  static id = 'NotImplementedError';
  static status = '501';
  static code = 'NotImplementedError';
  static title = 'NotImplementedError';
  static detail = '[NotImplementedError] - NotImplementedError';
}

export class BadGatewayError extends JsonApiError {
  static id = 'BadGatewayError';
  static status = '502';
  static code = 'BadGatewayError';
  static title = 'BadGatewayError';
  static detail = '[BadGatewayError] - BadGatewayError';
}

export class ServiceUnavailableError extends JsonApiError {
  static id = 'ServiceUnavailableError';
  static status = '503';
  static code = 'ServiceUnavailableError';
  static title = 'ServiceUnavailableError';
  static detail = '[ServiceUnavailableError] - ServiceUnavailableError';
}

export class GatewayTimeoutError extends JsonApiError {
  static id = 'GatewayTimeoutError';
  static status = '504';
  static code = 'GatewayTimeoutError';
  static title = 'GatewayTimeoutError';
  static detail = '[GatewayTimeoutError] - GatewayTimeoutError';
}

export { AggregateJsonApiError };

export default JsonApiError;
